// @zeainc/zea-web-components/dist/cjs/zea-images-gallery.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5b25f1f.js');

const zeaImagesGalleryCss = ":host{width:100%;height:100%}:host,input,button,select,textarea{font-family:'Roboto', sans-serif}.zea-ig-container{display:grid;width:100%;height:100%;grid-template-rows:repeat(5, 1fr);grid-template-columns:repeat(4, 1fr);gap:0.7em}.zea-ig-image.featured{grid-column:span 2;grid-row:span 2}.zea-ig-image{background-repeat:no-repeat;background-size:cover;background-position:center}.zea-ig-image:hover{filter:brightness(1.15);box-shadow:0px 0.25em 0.5em var(--color-shadow);z-index:1}.zea-ig-zoomed.inactive{display:none}.zea-ig-zoomed.active{display:block;position:fixed;background-color:var(--color-background-1);padding:1%;box-sizing:border-box;transition:all 0.5s;z-index:1000}.zea-ig-zoomed-inner{background-size:cover;background-position:center;background-repeat:no-repeat;width:100%;height:100%}.zea-ig-zoomed.active.animate .zea-ig-zoomed-inner{background-size:contain}";

const ZeaImagesGallery = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    /**
     * Number of featured images
     */
    this.featured = 2;
    /**
     * Number of columns
     */
    this.columns = 4;
    /**
     * Folder from which to take the files to display.
     */
    this.folder = {
      children: [],
    };
    this.zoomedImageClass = 'zea-ig-zoomed inactive';
    this.zoomedImageUrl = '';
  }
  /**
   * Number of featured images
   * @param {any} file The file object
   * @return {boolean} Whether or not the file extension is included in the imagesMediaTypes
   */
  static isFileImage(file) {
    return ZeaImagesGallery.imagesMediaTypes.includes(file.type);
  }
  /**
   * Take image back to original location
   */
  resetImageLocation() {
    this.zoomedTop = this.bbox.top + 'px';
    this.zoomedLeft = this.bbox.left + 'px';
    this.zoomedWidth = this.bbox.width + 'px';
    this.zoomedHeight = this.bbox.height + 'px';
    this.zoomedOpacity = '0';
  }
  /**
   * Zoom image in
   * @param {any} event The event
   * @param {any} image The image
   */
  zoomImageIn(event, image) {
    this.zoomedImageUrl = image.url;
    this.bbox = event.target.getBoundingClientRect();
    this.zoomedImageClass = 'zea-ig-zoomed active';
    this.resetImageLocation();
    setTimeout(() => {
      this.zoomedImageClass = 'zea-ig-zoomed active animate';
      this.zoomedTop = '0px';
      this.zoomedLeft = '0px';
      this.zoomedWidth = '100%';
      this.zoomedHeight = '100%';
      this.zoomedOpacity = '1';
    }, 100);
  }
  /**
   * Zoom image out
   */
  zoomImageOut() {
    this.resetImageLocation();
    setTimeout(() => {
      this.zoomedImageClass = 'zea-ig-zoomed inactive';
    }, 400);
  }
  /**
   * Main render function
   * @return {JSX} the generated html
   */
  render() {
    const onlyImages = Object.values(this.folder.children).filter(ZeaImagesGallery.isFileImage);
    const rowsCount = Math.ceil((onlyImages.length + this.featured * 3) / this.columns);
    console.log(onlyImages);
    return (index.h(index.Host, { class: "ZeaImagesGallery" }, index.h("div", { class: this.zoomedImageClass, style: {
        top: this.zoomedTop,
        left: this.zoomedLeft,
        width: this.zoomedWidth,
        height: this.zoomedHeight,
        opacity: this.zoomedOpacity,
      } }, index.h("div", { class: "zea-ig-zoomed-inner", style: {
        backgroundImage: 'url(' + this.zoomedImageUrl + ')',
      }, onClick: () => {
        this.zoomImageOut();
      } })), index.h("div", { class: "zea-ig-container", style: {
        gridTemplateColumns: 'repeat(' + this.columns + ', 1fr)',
        gridTemplateRows: 'repeat(' + rowsCount + ', 1fr)',
      } }, onlyImages.map((image, i) => (index.h("div", { style: {
        backgroundImage: `url(${image.url})`,
      }, class: `zea-ig-image ${i < this.featured ? ' featured' : ''}`, onClick: (event) => {
        this.zoomImageIn(event, image);
      }, title: image.name }))))));
  }
};
ZeaImagesGallery.imagesMediaTypes = ['image/jpeg', 'image/png', 'image/svg+xml'];
ZeaImagesGallery.style = zeaImagesGalleryCss;

exports.zea_images_gallery = ZeaImagesGallery;
