// @zeainc/zea-web-components/dist/cjs/zea-navigation-drawer.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5b25f1f.js');

const zeaNavigationDrawerCss = ":host,input,button,select,textarea{font-family:'Roboto', sans-serif}.zea-navigation-drawer{color:var(--color-foreground-1);display:inline-block;vertical-align:middle}.drawer{position:fixed;top:0;left:0;transform:translateX(-100%);width:192px;height:100vh;background-color:var(--color-background-1);z-index:10000000;transition:transform 0.3s;transition-timing-function:ease-out}.shown .drawer{left:0;transform:translateX(0);box-shadow:10px 3px 10px var(--color-background-4)}.toggle{position:relative;z-index:10000010;margin-right:12px;margin-left:7px;background-color:transparent;border-radius:50%;padding:4px;display:inline-block}.toggle:hover{background-color:var(--color-grey-3)}.drawer-content{padding-top:80px;padding-left:9px;font-size:14px}@media only screen and (max-width: 667px){.drawer{width:80%}}";

const ZeaNavigationDrawer = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.navDrawerOpen = index.createEvent(this, "navDrawerOpen", 7);
    this.navDrawerClosed = index.createEvent(this, "navDrawerClosed", 7);
    /**
     */
    this.shown = false;
  }
  /**
   * Listen to click events on the whole document
   * @param {any} e The event
   */
  handleClick(e) {
    if (!e.composedPath().includes(this.container) ||
      !e.composedPath().includes(this.toggleButton)) {
      this.shown = false;
      this.navDrawerClosed.emit(this);
    }
  }
  /**
   */
  onToggleClick() {
    this.shown = !this.shown;
    if (this.shown) {
      this.navDrawerOpen.emit(this);
    }
    else {
      this.navDrawerClosed.emit(this);
    }
  }
  /**
   */
  render() {
    return (index.h("div", { ref: (el) => (this.container = el), class: { 'zea-navigation-drawer': true, shown: this.shown } }, index.h("div", { class: "drawer" }, index.h("div", { class: "drawer-content" }, index.h("slot", null))), index.h("div", { class: "toggle", ref: (el) => (this.toggleButton = el), onClick: this.onToggleClick.bind(this) }, index.h("zea-icon", { size: 30, name: "menu" }))));
  }
};
ZeaNavigationDrawer.style = zeaNavigationDrawerCss;

exports.zea_navigation_drawer = ZeaNavigationDrawer;
