// @zeainc/zea-web-components/dist/cjs/zea-menu.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5b25f1f.js');

const zeaMenuCss = "*{cursor:default;user-select:none}:host{display:inline-block;vertical-align:middle}:host,input,button,select,textarea{font-family:'Roboto', sans-serif}.zea-menu{color:var(--color-foreground-1);flex-direction:column}.zea-menu .items{display:none}.menu-anchor{color:var(--color-foreground-1)}.zea-menu.shown{color:var(--color-foreground-1);display:flex;flex-direction:column}.zea-menu.shown .items{display:inline-flex;flex-direction:column}.dropdown slot{display:flex}.zea-menu.contextual{position:relative}.contextual.menu-wrap{display:inline-block;position:relative;vertical-align:middle}.contextual .items{position:absolute;z-index:1000}.contextual.show-anchor .items{margin-top:25px}.contextual.top-right .items{transform:translateX(-100%);margin-left:100%}.toolbar.horizontal .items{flex-direction:row}.toolbar{background-color:var(--color-background-2);position:fixed;border-radius:30px}.dropdown .items,.contextual .items{background-color:var(--color-background-2)}";

const ZeaMenu = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.offset = [0, 0];
    this.leftOffset = '';
    this.topOffset = '';
    this.type = 'dropdown';
    // for the toolbar type
    this.orientation = 'vertical';
    this.shown = false;
    this.showAnchor = false;
    this.anchorIcon = 'ellipsis-vertical-circle-outline';
    this.contextualAlign = 'top-left';
    this.anchorIconSize = 14;
  }
  /**
   * Listen to click (mouse up) events on the whole window
   * @param {any} ev the event
   */
  handleClick(ev) {
    if (ev.currentTarget == this.anchorElement) {
      this.shown = !this.shown;
      return;
    }
    else if (this.type == 'contextual' &&
      !this.isDescendant(this.hostElement, ev.target)) {
      this.shown = false;
    }
    this.mouseIsDown = false;
  }
  /**
   * isDescendant
   * @param {any} parent the parent
   * @param {any} child the parent
   * @return {any} whether or not is parent
   */
  isDescendant(parent, child) {
    let node = child.parentNode;
    while (node != null) {
      if (node == parent) {
        return true;
      }
      node = node.parentNode;
    }
    return false;
  }
  /**
   * Listen to zeaMenuItemClick events on the whole window
   * @param {any} ev The zeaMenuItemClick event
   */
  handleItemClick(ev) {
    if (this.type == 'contextual' && !ev.detail.hasSubitems) {
      this.shown = false;
    }
  }
  /**
   * Listen to mousedown event
   * @param {any} event the event
   */
  mousedownHandler(event) {
    if (this.type == 'toolbar' && event.currentTarget.tagName == 'ZEA-MENU') {
      this.mouseIsDown = true;
      this.offset = [
        this.menuElement.offsetLeft - event.clientX,
        this.menuElement.offsetTop - event.clientY,
      ];
    }
  }
  /**
   * Listen to mousemove event
   * @param {any} event the event
   */
  mousemoveHandler(event) {
    if (this.type == 'toolbar' && this.mouseIsDown) {
      this.menuElement.style.left = `${event.clientX + this.offset[0]}px`;
      this.menuElement.style.top = `${event.clientY + this.offset[1]}px`;
    }
  }
  /**
   * Listen to click events on anchor
   * @param {any} ev the event
   */
  handleAnchorClick(ev) {
    this.handleClick(ev);
    ev.stopPropagation();
  }
  /**
   * Called once when component first loads
   */
  componentDidLoad() {
    // the menu is hidden by default to avoid flashing
    if (this.type != 'contextual') {
      this.shown = true;
    }
  }
  /**
   * Called everytime the component renders
   * Apply the class to children
   */
  componentDidRender() {
    this.setupChildren();
    this.adjustPosition();
  }
  /**
   * Run some setup for the children items
   */
  setupChildren() {
    this.itemsContainer
      .querySelector('slot')
      .assignedNodes()
      .forEach((element) => {
      if ('type' in element) {
        element.type = this.type;
      }
      if ('itemParent' in element) {
        element.itemParent = this.hostElement;
      }
      if ('rootMenu' in element) {
        element.rootMenu = this.hostElement;
      }
    });
  }
  /**
   */
  adjustPosition() {
    if (this.type == 'contextual') {
      this.adjustContextualPosition();
    }
  }
  /**
   */
  adjustContextualPosition() {
    if (this.shown) {
      const winHeight = window.innerHeight;
      const winWidth = window.innerWidth;
      this.itemsContainer.style.transform = `translate(0, 0)`;
      const itemsBox = this.itemsContainer.getBoundingClientRect();
      const menuBox = this.menuElement.getBoundingClientRect();
      let trX = '0';
      let trY = '0';
      if (menuBox.left + itemsBox.width > winWidth) {
        trX = '-100%';
      }
      if (menuBox.top + itemsBox.height > winHeight) {
        trY = '-100%';
      }
      this.itemsContainer.style.transform = `translate(${trX}, ${trY})`;
    }
  }
  /**
   * Render function
   * @return {JSX}
   */
  render() {
    const anchor = this.type == 'contextual' && this.showAnchor ? (index.h("div", { onMouseUp: this.handleAnchorClick.bind(this), class: "menu-anchor", ref: (el) => (this.anchorElement = el) }, index.h("zea-icon", { size: this.anchorIconSize, name: this.anchorIcon }))) : null;
    const menuClass = {};
    menuClass['zea-menu'] = true;
    menuClass[this.type] = true;
    menuClass[this.orientation] = true;
    menuClass['shown'] = this.shown;
    menuClass[this.contextualAlign] = true;
    menuClass['show-anchor'] = this.showAnchor;
    const containerStyle = {
      top: this.topOffset,
      left: this.leftOffset,
    };
    return (index.h("div", { class: menuClass, ref: (el) => (this.menuElement = el) }, anchor, index.h("div", { ref: (el) => (this.itemsContainer = el), class: "items", style: containerStyle }, index.h("slot", null))));
  }
  get hostElement() { return index.getElement(this); }
};
ZeaMenu.style = zeaMenuCss;

exports.zea_menu = ZeaMenu;
