// @zeainc/zea-web-components/dist/cjs/zea-input-search.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-f5b25f1f.js');

const zeaInputSearchCss = ":host,input,button,select,textarea{font-family:'Roboto', sans-serif}.zea-input-search{color:var(--color-foreground-1);display:inline-flex;justify-content:flex-end;align-items:center;width:2em;border-radius:3px;transition:all 0.5s;background-color:transparent;border:1px solid transparent}path.icon{fill:var(--color-foreground-1)}.zea-input-search.active{width:100%;border:1px solid var(--color-grey-3);background-color:var(--color-background-1)}.zea-input-search-icon{height:2em;width:2em;display:inline-block;vertical-align:middle;border-radius:1em;box-sizing:border-box;padding:0.2em;flex-grow:0;flex-shrink:0}.zea-input-search-icon:hover{background-color:var(--color-grey-3);color:var(--color-button-text-1)}.zea-input-search.active .zea-input-search-icon:hover{background-color:var(--color-background-1)}input{display:block;height:100%;width:100%;border:none;background-color:transparent;outline:none;font-size:1em;color:var(--color-foreground-1)}.zea-input-search-text-container{height:2em;flex-grow:1;opacity:0;transition:all 0.5s}.zea-input-search.active .zea-input-search-text-container{opacity:1}.zea-input-cancel{display:none;height:1.5em;width:1.5em;border-radius:1em;box-sizing:border-box;flex-shrink:0}.active .zea-input-cancel{display:block;visibility:hidden}.tainted .zea-input-cancel{visibility:visible}";

const ZeaInputSearch = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.placeholder = '';
  }
  /**
   * Toggle 'active' class of the search box
   */
  toggleSearchBox() {
    this.mainContainer.classList.toggle('active');
    this.mainInput.focus();
  }
  /**
   */
  cancelSearch() {
    this.mainInput.value = '';
  }
  /**
   */
  componentDidLoad() {
    this.mainInput.addEventListener('keyup', () => {
      const q = this.mainInput.value.replace(/(^\s+|\s+$)/, '');
      if (q) {
        this.mainContainer.classList.add('tainted');
      }
      else {
        this.mainContainer.classList.remove('tainted');
      }
    });
  }
  /**
   * Main render function
   * @return {JSX} the generated html
   */
  render() {
    return (index.h("div", { class: "zea-input-search", ref: (el) => {
        this.mainContainer = el;
      } }, index.h("span", { class: "zea-input-search-icon", onClick: () => {
        this.toggleSearchBox();
      } }, index.h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "100%", height: "100%", viewBox: "0 0 24 24" }, index.h("path", { class: "icon", d: "M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z" }), index.h("path", { d: "M0 0h24v24H0z", fill: "none" }))), index.h("span", { class: "zea-input-search-text-container" }, index.h("input", { ref: (el) => {
        this.mainInput = el;
      }, onBlur: () => {
        setTimeout(() => {
          this.mainContainer.classList.remove('active');
        }, 100);
      }, onKeyDown: (e) => {
        e.stopPropagation();
      }, onKeyUp: (e) => {
        e.stopPropagation();
      }, class: "zea-input-search-text", type: "text", placeholder: this.placeholder })), index.h("span", { class: "zea-input-cancel", onClick: () => {
        this.cancelSearch();
      } }, index.h("svg", { xmlns: "http://www.w3.org/2000/svg", width: "100%", height: "100%", viewBox: "0 0 512 512" }, index.h("path", { d: "M289.94 256l95-95A24 24 0 00351 127l-95 95-95-95a24 24 0 00-34 34l95 95-95 95a24 24 0 1034 34l95-95 95 95a24 24 0 0034-34z", class: "icon" })))));
  }
  get mainElement() { return index.getElement(this); }
};
ZeaInputSearch.style = zeaInputSearchCss;

exports.zea_input_search = ZeaInputSearch;
