// @zeainc/zea-web-components/dist/esm/zea-param-widget-material-color.entry.js
import { r as registerInstance, h } from './index-b00b7e23.js';
import { u as uxFactory } from './UxFactory-aa8b881d.js';
import { i as iro$1 } from './iro.es-c9976626.js';

const zeaParamWidgetMaterialColorCss = ":host,input,button,select,textarea{font-family:'Roboto', sans-serif}.zea-param-widget-material-color{color:var(--color-foreground-1)}.color-sample{height:30px;line-height:30px;border:1px solid var(--color-foreground-1);margin-bottom:0.5em;text-align:center;font-size:0.8em}";

const ZeaParamWidgetMaterialColor = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.colorPickerHeight = 200;
    if (!globalThis.zeaEngine) {
      console.warn('`zeaEngine` missing from the `globalThis` object and this component requires it');
    }
  }
  /**
   * Listen to window resize event
   */
  handleWindowResize() {
    clearTimeout(this.resizeTimeout);
    this.resizeTimeout = setTimeout(() => this.resizeColorPicker(), 500);
  }
  /**
   * Resizes the  Color Picker
   */
  resizeColorPicker() {
    this.colorPicker.resize(this.container.offsetWidth, this.colorPickerHeight);
  }
  /**
   * Run when component loads
   */
  componentDidLoad() {
    this.setUpColorPicker();
    if (this.parameter) {
      this.parameter.on('valueChanged', () => {
        this.updateComponent();
      });
      this.updateComponent();
    }
  }
  /**
   * Called when the parameter value changes externally
   */
  updateComponent() {
    if (!this.parameterChange) {
      this.undoing = true;
      this.colorPicker.color.rgb = this.parameter.getValue().getAsRGBDict();
      this.undoing = false;
      this.sampleColor = this.colorPicker.color.hslString;
      this.setSampleTextColor();
    }
  }
  /**
   * Set the color of the text in the sample box
   */
  setSampleTextColor() {
    const l = 100 - this.colorPicker.color.hsl.l;
    this.sampleTextColor = `hsl(1, 0%, ${l}%)`;
  }
  /**
   * Setup the color picker and it's events
   */
  setUpColorPicker() {
    this.colorPicker = iro$1.ColorPicker(this.colorPickerContainer, {
      // Color picker options:
      // https://iro.js.org/guide.html
      width: this.container.offsetWidth,
      height: this.colorPickerHeight,
      anticlockwise: true,
      borderWidth: 0,
      borderColor: '#fff',
      sliderHeight: '10px',
      padding: 1,
      sliderMargin: 4,
      handleRadius: 4,
      layout: [
        {
          // hue slider
          component: iro$1.ui.Slider,
          options: {
            sliderType: 'hue',
          },
        },
        {
          // saturation slider
          component: iro$1.ui.Slider,
          options: {
            sliderType: 'saturation',
          },
        },
        {
          // regular value slider
          component: iro$1.ui.Slider,
          options: {},
        },
      ],
    });
    this.colorPicker.on('input:start', () => {
      this.handleUXChange(this.parameter.getValue());
    });
    this.colorPicker.on('input:end', () => {
      this.parameterChange = undefined;
    });
    this.colorPicker.on('color:change', () => {
      if (this.undoing)
        return;
      if (!globalThis.zeaEngine) {
        console.log('`zeaEngine` missing from `globalThis` object.');
        return;
      }
      const { Color } = globalThis.zeaEngine;
      const value = new Color();
      value.setFromRGBDict(this.colorPicker.color.rgb);
      this.sampleColor = this.colorPicker.color.hslString;
      this.setSampleTextColor();
      this.handleUXChange(value);
    });
  }
  /**
   * Handles the creation of the UX Change object and the UndoRedoManager
   *
   * @param {Parameter} value - Parameter object
   */
  handleUXChange(value) {
    if (!globalThis.zeaUx) {
      console.log('To add Undo/Redo functionalities just add `zeaUx` library to your `globalThis` object, because right now it is missing.');
      return;
    }
    const { UndoRedoManager, ParameterValueChange } = globalThis.zeaUx;
    const undoRedoManager = UndoRedoManager.getInstance();
    if (!this.parameterChange) {
      this.parameterChange = new ParameterValueChange(this.parameter, value);
      undoRedoManager.addChange(this.parameterChange);
    }
    else {
      this.parameterChange.update({ value });
    }
  }
  /**
   * Main render method.
   * @return {JSX} The generated html
   */
  render() {
    clearTimeout(this.resizeTimeout);
    this.resizeTimeout = setTimeout(() => this.resizeColorPicker(), 500);
    return (h("div", { ref: (el) => (this.container = el), class: "zea-param-widget-material-color" }, h("div", { style: {
        color: this.sampleTextColor,
        backgroundColor: this.sampleColor,
      }, class: "color-sample" }, this.sampleColor), h("div", { ref: (el) => (this.colorPickerContainer = el), class: "color-picker" })));
  }
};
uxFactory.registerWidget('zea-param-widget-material-color', (p) => p.getDataType() === 'Color' && 'getImage' in p);
ZeaParamWidgetMaterialColor.style = zeaParamWidgetMaterialColorCss;

export { ZeaParamWidgetMaterialColor as zea_param_widget_material_color };
