// @zeainc/zea-web-components/dist/collection/components/zea-param-widget-number/zea-param-widget-number.js
import { Component, h, Prop, State, Watch } from '@stencil/core';
import uxFactory from '../../assets/UxFactory.js';
/**
 * Main class for the component
 */
export class ZeaParamWidgetNumber {
  /**
   * Class constructor
   */
  constructor() {
    this.onChange = this.onChange.bind(this);
    if (!globalThis.zeaEngine) {
      console.warn('`zeaEngine` missing from the `globalThis` object and this component requires it');
    }
  }
  /**
   * Re-init input when parameter changes
   */
  parameterChangeHandler() {
    this.setUpInput();
  }
  /**
   * Run when component loads
   */
  componentDidLoad() {
    if (this.parameter) {
      this.setUpInput();
    }
  }
  /**
   * Set up the input
   */
  setUpInput() {
    this.range = this.parameter.getRange();
    this.step = this.parameter.getStep();
    this.updateComponent();
    this.parameter.on('valueChanged', () => {
      this.updateComponent();
    });
  }
  /**
   * Sets the value of the input
   */
  updateComponent() {
    if (this.range) {
      this.value =
        ((this.parameter.getValue() - this.range[0]) /
          (this.range[1] - this.range[0])) *
          200;
    }
    else {
      this.value = this.parameter.getValue();
    }
  }
  /**
   * Run when input changes
   */
  onChange() {
    let value = this.round(this.inputField.valueAsNumber);
    if (this.range) {
      // Renmap from the 0..200 integer to the floating point
      // range specified in the parameter.
      value = this.range[0] + (value / 200) * (this.range[1] - this.range[0]);
      value = this.clamp(value, this.range[0], this.range[1]);
    }
    this.handleUXChange(value);
  }
  /**
   * Handles the creation of the UX Change object and the UndoRedoManager
   *
   * @param {Parameter} value - Parameter object
   */
  handleUXChange(value) {
    if (!globalThis.zeaUx) {
      this.parameter.setValue(value);
      console.log('To add Undo/Redo functionalities just add `zeaUx` library to your `globalThis` object, because right now it is missing.');
      return;
    }
    const { UndoRedoManager, ParameterValueChange } = globalThis.zeaUx;
    const undoRedoManager = UndoRedoManager.getInstance();
    const change = new ParameterValueChange(this.parameter, value);
    undoRedoManager.addChange(change);
  }
  /**
   * Round number
   * @param {number} value Number to be rounded
   * @param {number} decimals Number of decimal places
   * @return {number} Rounded number
   */
  round(value, decimals = 6) {
    return Number(Math.round(Number(value + 'e' + decimals)) + 'e-' + decimals);
  }
  /**
   * Clamp number
   * @param {number} num Number to be rounded
   * @param {number} a Number of decimal places
   * @param {number} b Number of decimal places
   * @return {number} clamped number
   */
  clamp(num, a, b) {
    return Math.max(Math.min(num, Math.max(a, b)), Math.min(a, b));
  }
  /**
   * Render method.
   * @return {JSX} The generated html
   */
  render() {
    if (this.range) {
      return (h("div", { class: "zea-param-widget-number" },
        h("input", { onChange: this.onChange, ref: (el) => (this.inputField = el), class: "mdl-slider mdl-js-slider", type: "range", min: "0", max: "200", step: this.step ? this.step : 1, id: this.parameter.getName(), value: this.value, tabindex: "0" })));
    }
    return (h("div", { class: "zea-param-widget-number" },
      h("input", { onChange: this.onChange, ref: (el) => (this.inputField = el), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", id: this.parameter.getName(), value: this.value, tabindex: "0" })));
  }
  static get is() { return "zea-param-widget-number"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["zea-param-widget-number.css"]
  }; }
  static get styleUrls() { return {
    "$": ["zea-param-widget-number.css"]
  }; }
  static get properties() { return {
    "parameter": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Parameter to be edited"
      },
      "attribute": "parameter",
      "reflect": false
    },
    "inputType": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Html input type"
      },
      "attribute": "input-type",
      "reflect": false
    },
    "appData": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "The application data"
      },
      "attribute": "app-data",
      "reflect": false
    },
    "value": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "The value of the number"
      },
      "attribute": "value",
      "reflect": false
    }
  }; }
  static get states() { return {
    "range": {},
    "step": {}
  }; }
  static get watchers() { return [{
      "propName": "parameter",
      "methodName": "parameterChangeHandler"
    }]; }
}
uxFactory.registerWidget('zea-param-widget-number', (p) => p.getDataType() === 'Number');
