// @zeainc/zea-web-components/dist/collection/components/zea-param-widget-xfo/zea-param-widget-xfo.js
/* eslint-disable no-unused-vars */
import { Component, h, Prop } from '@stencil/core';
import uxFactory from '../../assets/UxFactory.js';
/**
 * Main class for the component
 */
export class ZeaParamWidgetXfo {
  /**
   * Class constructor
   */
  constructor() {
    this.settingValue = false;
    this.onInput = this.onInput.bind(this);
    this.onBlur = this.onBlur.bind(this);
    if (!globalThis.zeaEngine) {
      console.warn('`zeaEngine` missing from the `globalThis` object and this component requires it');
    }
  }
  /**
   * Run when component loads
   */
  componentDidLoad() {
    if (this.parameter) {
      this.setUpInputs();
      this.updateComponent();
    }
  }
  /**
   * Value change handler
   */
  updateComponent() {
    if (!this.settingValue) {
      const xfo = this.parameter.getValue();
      this.trxField.value = `${this.round(xfo.tr.x)}`;
      this.tryField.value = `${this.round(xfo.tr.y)}`;
      this.trzField.value = `${this.round(xfo.tr.z)}`;
      this.orxField.value = `${this.round(xfo.ori.x)}`;
      this.oryField.value = `${this.round(xfo.ori.y)}`;
      this.orzField.value = `${this.round(xfo.ori.z)}`;
      this.orwField.value = `${this.round(xfo.ori.w)}`;
      this.scxField.value = `${this.round(xfo.sc.x)}`;
      this.scyField.value = `${this.round(xfo.sc.y)}`;
      this.sczField.value = `${this.round(xfo.sc.z)}`;
    }
  }
  /**
   * Set the inputs up
   */
  setUpInputs() {
    this.parameter.on('valueChanged', () => {
      this.updateComponent();
    });
  }
  /**
   * Input handler
   */
  onInput() {
    this.settingValue = true;
    if (!globalThis.zeaEngine) {
      console.log('`zeaEngine` missing from `globalThis` object.');
      return;
    }
    const { Xfo } = globalThis.zeaEngine;
    const value = new Xfo();
    value.tr.set(this.trxField.valueAsNumber, this.tryField.valueAsNumber, this.trzField.valueAsNumber);
    value.ori.set(this.orxField.valueAsNumber, this.oryField.valueAsNumber, this.orzField.valueAsNumber, this.orwField.valueAsNumber); /* value order is xyzw*/
    value.ori.normalizeInPlace();
    value.sc.set(this.scxField.valueAsNumber, this.scyField.valueAsNumber, this.sczField.valueAsNumber);
    this.handleUXChange(value);
  }
  /**
   * Change handler
   */
  onBlur() {
    this.settingValue = false;
    this.parameterChange = undefined;
  }
  /**
   * Handles the creation of the UX Change object and the UndoRedoManager
   *
   * @param {Parameter} value - Parameter object
   */
  handleUXChange(value) {
    if (!globalThis.zeaUx) {
      console.log('To add Undo/Redo functionalities just add `zeaUx` library to your `globalThis` object, because right now it is missing.');
      return;
    }
    const { UndoRedoManager, ParameterValueChange } = globalThis.zeaUx;
    const undoRedoManager = UndoRedoManager.getInstance();
    if (!this.parameterChange) {
      this.parameterChange = new ParameterValueChange(this.parameter, value);
      undoRedoManager.addChange(this.parameterChange);
    }
    else {
      this.parameterChange.update({ value });
    }
  }
  /**
   * Round number
   * @param {number} value the value to be rounded
   * @param {number} decimals decimal places to keep
   * @return {number} the rouunded value
   */
  round(value, decimals = 6) {
    if (Math.abs(value) < Number('1e-6'))
      return 0;
    return Number(Math.round(Number(value + 'e' + decimals)) + 'e-' + decimals);
  }
  /**
   * Render method.
   * @return {JSX} The generated html
   */
  render() {
    return (h("div", { class: "zea-param-widget-xfo" },
      h("fieldset", null,
        h("legend", null, "Translation"),
        h("div", { class: "input-wrap" },
          h("label", null, "X"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.trxField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "Y"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.tryField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "Z"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.trzField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" }))),
      h("fieldset", null,
        h("legend", null, "Rotation"),
        h("div", { class: "input-wrap" },
          h("label", null, "X"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.orxField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "Y"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.oryField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "Z"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.orzField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "W"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.orwField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" }))),
      h("fieldset", null,
        h("legend", null, "Scale"),
        h("div", { class: "input-wrap" },
          h("label", null, "X"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.scxField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "Y"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.scyField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })),
        h("div", { class: "input-wrap" },
          h("label", null, "Z"),
          h("input", { onInput: this.onInput, onBlur: this.onBlur, ref: (el) => (this.sczField = el), id: this.parameter.getName(), type: "number", pattern: "-?[0-9]*(.[0-9]+)?", tabindex: "0" })))));
  }
  static get is() { return "zea-param-widget-xfo"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["zea-param-widget-xfo.css"]
  }; }
  static get styleUrls() { return {
    "$": ["zea-param-widget-xfo.css"]
  }; }
  static get properties() { return {
    "parameter": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Parameter to be edited"
      },
      "attribute": "parameter",
      "reflect": false
    },
    "appData": {
      "type": "any",
      "mutable": false,
      "complexType": {
        "original": "any",
        "resolved": "any",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "The application data"
      },
      "attribute": "app-data",
      "reflect": false
    }
  }; }
}
uxFactory.registerWidget('zea-param-widget-xfo', (p) => p.getDataType() === 'Xfo');
